### PR TITLE
Virtual layers

### PR DESCRIPTION
**Pull request by mhugo**
Not for merging yet. For preliminary reviews, feedback, if available :)

This PR adds a new vector provider for virtual layers, following this [QEP](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/44).

Compared to the previous prototype implementation, it has been cleaned up. The main difference is that no flex/bison parser is needed anymore. I find it was too much code and hard to make sure the grammar was exactly the one used by the current SQLite engine (considering we are linked to an external sqlite lib)
The current strategy to detect referenced tables and column types (including the geometry) in a query is then:
- use SQLite to introspect what is possible to introspect
- allow the user to specify things either in the URL of the construction (fields= and geometry= options) or by using special comments in the SQL query (SELECT t /&ast;:int&ast;/, st_makepoint(geometry) as geom /&ast;:point:4326&ast;/ FROM ...)
- get types from a  SELECT ... LIMIT 1 at last resort

This patch includes also the DB manager plugin for virtual layers.

What is missing:
- the unit tests segfault once in a while. Not sure yet what is the problem
- virtual layers that use other loaded layers as primary sources must be reloaded _after_ their dependencies when the project is opened. I may have to find a way to declare dependencies between layers somehow.
- the main UI is the DB manager plugin. I still have to cleanup some code (and UI) for a smoother integration into the main app

### CODE REVIEW

File: python/plugins/db_manager/dlg_sql_window.py [hunk 1]
```diff
@@ -91,10 +91,8 @@ def __init__(self, iface, db, parent=None):
         self.updatePresetsCombobox()
 
         self.geomCombo.setEditable(True)
-        self.geomCombo.lineEdit().setReadOnly(True)
 
         self.uniqueCombo.setEditable(True)
-        self.uniqueCombo.lineEdit().setReadOnly(True)
```
Comment (SebDieBln) [inline python/plugins/db_manager/dlg_sql_window.py]: Under what circumstances would the user want to enter a value that is not a column in ```geomCombo```or ```uniqueCombo```? Or actually a list of columns for the latter, if the provider supports composite keys (currently only Postgres does).

**Comment by SebDieBln** [comment]
:+1:, great feature!

**Comment by NathanW2** [comment]
> the main difference is that no flex/bison parser is needed anymore.

Great move.

Good work.  Going to give it a try now.

What ability do we have to increase query speed if that is a issue in the
future?

File: tests/src/python/test_provider_virtual.py [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+    print "You should install pyspatialite to run the tests"
+    raise ImportError
+
+import tempfile
+
+# Convenience instances in case you may need them
+QGISAPP, CANVAS, IFACE, PARENT = getQgisTestApp()
+
+
+class TestQgsVirtualLayerProvider(TestCase):
```
Comment (nyalldawson) [inline tests/src/python/test_provider_virtual.py]: Can I suggest making this

    class TestQgsVirtualLayerProvider(TestCase, ProviderTestCase):

And adding the provider test data from `tests/testdata/provider` too? ProviderTestCase is designed to ensure consistent behaviour across all vector data providers, so will be good to identify any potential issues with this provider

**Comment by NathanW2** [comment]
I get this trying to build

`/home/nathan/dev/qgis/src/providers/virtual/qgsvirtuallayersqlitehelper.cpp:137: error: 'runtime_error' is not a member of 'std'
       throw std::runtime_error( sqlite3_errmsg( db_ ) );
             ^
`

File: python/plugins/db_manager/dlg_sql_window.py (hunk 1 above)
Comment (mhugo) [inline python/plugins/db_manager/dlg_sql_window.py]: When the provider is not able to detect the name of the geometry / unique id column. This might be the case for virtual layers in rare circumstances, for example when the query contains more than one geometry column

File: tests/src/python/test_provider_virtual.py (hunk 2 above)
Comment (mhugo) [inline tests/src/python/test_provider_virtual.py]: Sure

**Comment by mhugo** [comment]
@nyalldawson Travis complains about a missing SIP binding : QgsVirtualLayerDefinition::SourceLayer.SourceLayer
But it exists ... is there something specific to do ?

**Comment by nyalldawson** [comment]
@mhugo hmm... not sure on that one. It may be that the test can't handle the nested classes correctly. I'd say either increase the coverage threshold to cover this, or add some missing bindings from elsewhere to keep Travis happy. I'll try and debug when this is merged into master.

**Comment by mhugo** [comment]
@nyalldawson yes apparently it was the test which adds constructors of nested classes as missing members. the last commit has a possible fix

Le 15 décembre 2015 04:39:52 GMT+01:00, Nyall Dawson <notifications@github.com> a écrit :
>@mhugo hmm... not sure on that one. It may be that the test can't
>handle the nested classes correctly. I'd say either increase the
>coverage threshold to cover this, or add some missing bindings from
>elsewhere to keep Travis happy. I'll try and debug when this is merged
>into master.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/qgis/QGIS/pull/2568#issuecomment-164635791

-- 
Envoyé de mon appareil Android avec K-9 Mail. Veuillez excuser ma brièveté.

File: python/plugins/db_manager/dlg_sql_window.py (hunk 1 above)
Comment (SebDieBln) [inline python/plugins/db_manager/dlg_sql_window.py]: Then there is a misunderstanding:
Your change allows the user to specify __arbitrary values__ for geometry / unique id. The comboboxes are actually there to force the user to choose a valid column name, regardless of the type.

So unless there is the need for arbitrary values, e.g. because the provider can not determine the column names at all, please consider reverting these changes.

Comment (mhugo) [inline python/plugins/db_manager/dlg_sql_window.py]: The virtual layer provider is not always able to detect all the column names. Then sometimes, the user may want to specify them by hand. The comboboxes are still there for suggestions, if the suggestions can be detected.

**Comment by mhugo** [comment]
So now the status is better: no more segfault (including under OSX) and the provider base tests also allowed to fix a some bugs :)

File: python/plugins/db_manager/dlg_sql_window.py (hunk 1 above)
Comment (SebDieBln) [inline python/plugins/db_manager/dlg_sql_window.py]: > The virtual layer provider is not always able to detect all the column names.

I am almost sure this will lead to serious problems :disappointed: 
> the main UI is the DB manager plugin. I still have to cleanup some code (and UI) for a smoother integration into the main app

:+1:, so this modification of the DB Manager will not be necessary then at all.

Comment (mhugo) [inline python/plugins/db_manager/dlg_sql_window.py]: I am not sure to measure the importance of such a modification ... is it really a problem to let the user enter its own column name if he is not happy with the column names given by the combo box? What should possibly happen ? A wrong column name ? Well, it is up to the user to know what he types ...

Comment (SebDieBln) [inline python/plugins/db_manager/dlg_sql_window.py]: The impact of this modification is rather small, because it does not break anything. It's just a slight degradation of the UI if the user can provoke an error the software could have prevented beforehand.
It can be minimized further by only applying the modification when your new provider is actually used and not for the other providers. I did a similar thing when implementing composite keys in DB-Manager for the Postgres-provider, see #2382. Although the solution there is more a hack than a good piece of software engineering.

I think the bigger issue is, why can the provider not determine all column names? The way I understand the abstraction level in QGIS the provider is responsible for defining the columns and any column it can not detect is just non-existing to the rest of the system.

If you could sketch such a situation I'd be happy to have a look at the code.

**Comment by m-kuhn** [comment]
I haven't had the time to look into the code. But I get to like the idea :)
Sorry for asking for more, but do you have an estimate how hard it would be to add update support?

File: python/plugins/db_manager/dlg_sql_window.py (hunk 1 above)
Comment (mhugo) [inline python/plugins/db_manager/dlg_sql_window.py]: After reconsideration, I think it is safe to revert this change. The provider is actually able to detect column names, except in a very specific situation where more than one geometry column is present in the query, but anyway qgis supports only one geometry column ...

**Comment by mhugo** [comment]
@m-kuhn Adding an update mode would need:
- to implement the xUpdate() function in the sqlite module: I think it is nothing more than calls to QgsVectorLayer::changeValues / addFeatures / etc.
- allow the user to add SQLite trigger codes for the view representing the query
Not much code, but it has to be tested carefully.

**Comment by mhugo** [comment]
@NathanW2 Sorry I forgot to answer your question. If performances become an issue, it is hard to tell now where it could come from, but I guess that could be related to the conversion between QgsGeometry and the spatialite binary form. They are quite similar but still require some conversion (and allocation / copy). So ... I am thinking about a way to "pass" native qgsgeometry without copy, and converting them only when a spatialite function is called. Another option would be to only rely on QgsExpression functions and forget spatialite functions ... not sure yet

**Comment by mhugo** [comment]
I merge, since it is already usable. I will still be working on it for the 2.14

**Comment by SebDieBln** [comment]
Is there a reason why the files
* src/core/qgsvirtuallayerdefinition.cpp
* src/core/qgsvirtuallayerdefinition.h

are in _src/core/_ and the classes `CORE_EXPORT`ed? Seems to me like they are only used by _providers/virtual_?

**Comment by mhugo** [comment]
@SebDieBln Theses classes are exposed to the Python API, in order to ease the manipulation of source URI for virtual layers. It follows the same idea as QgsDataSourceURI for rdbms providers
